scripts/s2e-build-image.sh

```diff
@@ -0,0 +1,586 @@
+    mkfs.ext4 -F ${NBD_DEV}p2
+}
+
+prepare_image() {
+    log "Preparing the image"; {
+    if ! test -f ${QEMUIMG}; then
+        qemu-img create -f raw ${QEMUIMG}  ${QCOW_SIZE}G
+        chown ${SUDO_USER}:${SUDO_USER} ${QEMUIMG}
+    else
+        warn "The image ${QEMUIMG} already exist"; {}
+    fi
+    }
+}
+
+sync_rootfs_image() {
+    log "Installing rootfs in the qcow image"; {
+        rsync -az --exclude "root/.cache" --partial ${ROOTFS}/ ${MNTDIR}
+    }
+}
+
+sync_s2e_module() {
+    log "Installing s2e kprobe kernel module"; {
+        rsync -L -az --exclude ".git/" --partial ${WORKDIR}/s2e-kprobe ${ROOTFS}/root/
+    }
+}
+
+install_users2e_files() {
+    log "Installing override file for autologin"; {
+        mkdir -p ${ROOTFS}/home/s2e/ 2> /dev/null || true
+        mkdir -p ${ROOTFS}/etc/systemd/system/getty@tty1.service.d/ 2> /dev/null || true
+        mkdir -p ${ROOTFS}/etc/systemd/system/getty@ttyS0.service.d/ 2> /dev/null || true
+        cp ${OVERRIDE} ${ROOTFS}/etc/systemd/system/getty@tty1.service.d/
+        cp ${OVERRIDE} ${ROOTFS}/etc/systemd/system/getty@ttyS0.service.d/
+        rsync -az ${GUESTTOOLS}/ ${ROOTFS}/home/s2e/
+        cp ${S2EHOME}/.bash_login ${ROOTFS}/home/s2e/
+    }
+}
+
+init_workdir () {
+    log "Initializing working directory at ${WORKDIR}"; {
+        mkdir -p ${WORKDIR}  2> /dev/null || true
+        chown ${SUDO_USER}:${SUDO_USER} ${WORKDIR}
+    }
+    log "Initializing kernel directory at ${KERNELPATH}"; {
+        mkdir -p ${KERNELPATH} 2> /dev/null || true
+        chown ${SUDO_USER}:${SUDO_USER} ${KERNELPATH}
+    }
+    log "Initializing output directory at ${OUTPUTDIR}"; {
+        mkdir -p ${OUTPUTDIR} 2> /dev/null || true
+    }
+    log "Initializing mount directory ${MNTDIR}"; {
+        mkdir -p ${MNTDIR} 2> /dev/null || true
+    }
+}
+
+generate_final_s2e_image () {
+    log "Creating s2e image ${QEMUIMGS2E} from ${QEMUIMG}"; {
+        cp --reflink=auto ${QEMUIMG} ${QEMUIMGS2E}
+        chown ${SUDO_USER}:${SUDO_USER} ${QEMUIMGS2E}
+    }
+}
+
+create_snapshot () {
+    log "Create snapshot"; {
+        LD_PRELOAD=${S2EDIR}/install/share/libs2e/libs2e-x86_64.so ${S2EDIR}/install/bin/qemu-system-x86_64 -enable-kvm -drive if=ide,index=0,file=${QEMUIMGS2E},format=s2e,cache=writeback -serial file:${OUTPUTDIR}/serial_ready.txt -enable-serial-commands -net none -net nic,model=e1000 -m 1G -nographic
+        chown ${SUDO_USER}:${SUDO_USER} ${QEMUIMGS2E}.ready
+    }
+}
+
+create_manifest() {
+    log "Create manifest"; {
+        ${GUESTIMG}/scripts/generate_image_descriptor.py -i ${GUESTIMG}/images.json -o ${OUTPUTDIR}/image.json -n debootstrap snapshot="ready" qemu_build="x86_64" memory="1G" qemu_extra_flags="-net none -net nic,model=e1000"
+    }
+}
+
+copy_guestfs() {
+    log "Creating/Updating guestfs files in output"; {
+        mkdir -p ${OUTPUTDIR}/guestfs 2> /dev/null || true
+        rsync -az --partial ${ROOTFS}/ ${OUTPUTDIR}/guestfs/
+    }
+}
+
+main() {
```

Comment:
What parameters does this script take? Some comment explaining how it's invoked would help.

Reply:
Ah I see them above. It would be good to add some default invocation example to build a complete image.

Reply:
It would be good to integrate it into s2e image_build, perhaps separately from the existing Makefile-based system. The feature should be well visible to users.